torch/csrc/distributed/c10d/init.cpp

```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
           "get_future",
           [](::c10d::ProcessGroup::Work& work)
               -> std::shared_ptr<jit::PythonFutureWrapper> {
+#ifdef USE_C10D_NCCL
+            auto fut = dynamic_cast<::c10d::ProcessGroupNCCL::FutureNCCL*>(
+                work.getFuture().get());
+            // By default record_stream_cb_ of the future is not set, and here
+            // we assume the type of this future value is or can be casted into
+            // a tensor vector that only has exactly one tensor.
+            if (!fut->hasRecordStreamCallback()) {
+              auto record_stream_cb = [](const at::IValue& value,
+                                         at::cuda::CUDAStream cuda_stream) {
+                if (value.isTensorList() || value.isPyObject()) {
+                  pybind11::gil_scoped_acquire gil;
```

Comment:
I think this function already holds GIL (because there is no gil_scoped_release guard in the definition)? You can probably check this by using `PyGILState_Check`. If so, I think re-acquiring the GIL here could result in some issues.

Reply:
The new design moved this part to pybind_utils.h